File: src/NubArca.Api/Domain/PartyMessageText.cs
using System.Globalization;
using System.Text;

namespace NubArca.Api.Domain;

// The ONE definition of what a party message's text is, and of how long it is.
//
// Both halves are a CONTRACT shared with the browser, not an implementation
// detail, because the guest sees a live character counter and the server is the
// authority: if the two disagree, a guest watches the counter say 118 and the
// submit fail. `frontend/src/lib/partyMessageText.ts` is the mirror of this
// file, expressed in the JavaScript constructs that match these rules exactly,
// and PartyMessageTextTests pins the shared cases.
//
// LENGTH IS COUNTED IN UNICODE CODE POINTS. Not UTF-16 units, which would
// charge two for every emoji and vary with an implementation detail of the
// storage encoding; and not grapheme clusters, which are the friendliest count
// but are defined by an ICU table that .NET and a browser upgrade on their own
// schedules — the day they disagree, the counter and the validator disagree
// too. Code points are exactly `EnumerateRunes().Count()` here and exactly
// `[...text].length` there, on every runtime, forever. The visible consequence
// is that a heart with a variation selector costs 2 and a family emoji costs 7;
// with a 120 budget for a party greeting that is a price worth paying for two
// numbers that can never drift apart.
public static class PartyMessageText
{
    // A message is plain text on one line. Normalisation is applied BEFORE
    // validation, so the limit is measured against what will actually be
    // stored and displayed — never against whitespace the guest cannot see.
    //
    //   1. Drop format characters (Cf). This is what removes the bidi overrides
    //      (U+202A-U+202E, U+2066-U+2069) that can make stored text render as
    //      something other than what it says, plus zero-width spaces and soft
    //      hyphens used to pad past a limit. ZWJ and ZWNJ are KEPT: they are
    //      what holds an emoji sequence and much Indic/Persian text together,
    //      and dropping them would corrupt real messages.
    //   2. Turn every control character and every whitespace character —
    //      including the line endings, which is how CRLF/CR/LF all become the
    //      same thing — into a plain space.
    //   3. Collapse runs of spaces, then trim.
    public static string Normalize(string? value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return string.Empty;
        }

        var builder = new StringBuilder(value.Length);
        var pendingSpace = false;

        foreach (var rune in value.EnumerateRunes())
        {
            var category = Rune.GetUnicodeCategory(rune);

            if (category == UnicodeCategory.Format && !IsJoiner(rune))
            {
                continue;
            }

            if (category == UnicodeCategory.Control || Rune.IsWhiteSpace(rune))
            {
                // Never emitted at the start, and only emitted once a further
                // non-space rune arrives — which collapses runs and trims both
                // ends in a single pass.
                pendingSpace = builder.Length > 0;
                continue;
            }

            if (pendingSpace)
            {
                builder.Append(' ');
                pendingSpace = false;
            }

            builder.Append(rune);
        }

        return builder.ToString();
    }

    // Length in Unicode code points — the counted unit for BOTH limits.
    public static int Length(string? value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return 0;
        }

        var count = 0;
        foreach (var _ in value.EnumerateRunes())
        {
            count++;
        }
        return count;
    }

    // The optional guest signature: normalised, then absent rather than empty.
    // "" and "   " and null all become null, so the TV has exactly one shape to
    // render for "this guest did not sign it".
    //
    // Returns false when the name is present but too long — a name over the
    // limit is a rejected submission, never a silent truncation of what
    // somebody chose to call themselves.
    public static bool TryNormalizeDisplayName(string? value, out string? normalized)
    {
        var text = Normalize(value);
        if (text.Length == 0)
        {
            normalized = null;
            return true;
        }

        if (Length(text) > PartyMessageLimits.MaxDisplayNameLength)
        {
            normalized = null;
            return false;
        }

        normalized = text;
        return true;
    }

    // The message body: normalised, non-empty, within the limit. A body that is
    // empty or whitespace-only after normalisation is rejected — there is no
    // such thing as a blank greeting on a television.
    public static bool TryNormalizeBody(string? value, out string normalized)
    {
        normalized = Normalize(value);
        if (normalized.Length == 0)
        {
            return false;
        }

        return Length(normalized) <= PartyMessageLimits.MaxBodyLength;
    }

    // U+200C ZERO WIDTH NON-JOINER and U+200D ZERO WIDTH JOINER.
    private static bool IsJoiner(Rune rune) => rune.Value is 0x200C or 0x200D;
}

File: frontend/packages/api-client/src/partyMessageText.ts
// The browser half of the party-message text contract. Its C# mirror is
// `src/NubArca.Api/Domain/PartyMessageText.cs`, which is the AUTHORITY: this
// file exists so the guest's live character counter says the same number the
// server is about to measure. If the two drift, a guest watches the counter say
// "2 left" and the submit fail, with nothing on screen to explain it.
//
// LENGTH IS COUNTED IN UNICODE CODE POINTS — `[...text].length` here, and
// `EnumerateRunes().Count()` there. Not UTF-16 units (`text.length`), which
// would charge two for every emoji; and not grapheme clusters, which are the
// friendliest count but come from an ICU table that a browser and .NET upgrade
// on their own schedules. Code points are the same number on every runtime.

export const PARTY_MESSAGE_LIMITS = {
  displayName: 40,
  text: 120,
} as const;

// Matches PartyMessageText.Normalize step for step:
//   1. drop format characters (Cf) — bidi overrides, zero-width padding, soft
//      hyphens — but KEEP the two joiners, which hold emoji sequences and much
//      Indic/Persian text together;
//   2. turn every control character and every whitespace character (all the
//      line endings included) into a plain space;
//   3. collapse runs of spaces, then trim.
export function normalizePartyMessageText(value: string | null | undefined): string {
  if (!value) return '';
  return value
    .replace(/\p{Cf}/gu, (c) => (c === '‌' || c === '‍' ? c : ''))
    .replace(/[\p{Cc}\p{White_Space}]/gu, ' ')
    .replace(/ {2,}/g, ' ')
    .trim();
}

// Length in Unicode code points. The spread is deliberate: `value.length` would
// count UTF-16 units and disagree with the server on the first emoji typed.
export function partyMessageLength(value: string | null | undefined): number {
  return value ? [...value].length : 0;
}

// What the counter under the textarea should show, measured the way the server
// will measure it — so a guest padding with spaces sees the number stop moving
// rather than watching it climb into a rejection.
export function partyMessageRemaining(value: string | null | undefined): number {
  return PARTY_MESSAGE_LIMITS.text - partyMessageLength(normalizePartyMessageText(value));
}

export function partyDisplayNameRemaining(value: string | null | undefined): number {
  return PARTY_MESSAGE_LIMITS.displayName - partyMessageLength(normalizePartyMessageText(value));
}

// The client-side answer to "may this be submitted". The server re-decides the
// same question from the same rules; this only spares the guest a round-trip
// and lets the button disable itself.
export function isPartyMessageSubmittable(
  text: string | null | undefined,
  displayName: string | null | undefined,
): boolean {
  const body = normalizePartyMessageText(text);
  if (body.length === 0) return false;
  if (partyMessageLength(body) > PARTY_MESSAGE_LIMITS.text) return false;
  return partyMessageLength(normalizePartyMessageText(displayName)) <= PARTY_MESSAGE_LIMITS.displayName;
}

// ── The guest book ──────────────────────────────────────────────────────────
//
// The SAME normalisation, with the book's own limits. A dedication is written
// to be kept rather than read out over music, so it gets more room — but the
// rules about what text IS do not change, and there is deliberately no second
// implementation of them here: `normalizePartyMessageText` is the one mirror of
// `PartyMessageText.Normalize`, and the backend's `PartyGuestbookText` reuses
// the same function on its side for the same reason.

export const PARTY_GUESTBOOK_TEXT_LIMITS = {
  authorDisplayName: 80,
  body: 1000,
} as const;

export function partyGuestbookBodyRemaining(value: string | null | undefined): number {
  return PARTY_GUESTBOOK_TEXT_LIMITS.body
    - partyMessageLength(normalizePartyMessageText(value));
}

export function partyGuestbookAuthorRemaining(value: string | null | undefined): number {
  return PARTY_GUESTBOOK_TEXT_LIMITS.authorDisplayName
    - partyMessageLength(normalizePartyMessageText(value));
}

export function isPartyGuestbookSubmittable(
  body: string | null | undefined,
  authorDisplayName: string | null | undefined,
): boolean {
  const text = normalizePartyMessageText(body);
  if (text.length === 0) return false;
  if (partyMessageLength(text) > PARTY_GUESTBOOK_TEXT_LIMITS.body) return false;
  return partyMessageLength(normalizePartyMessageText(authorDisplayName))
    <= PARTY_GUESTBOOK_TEXT_LIMITS.authorDisplayName;
}
